go.mod

```diff
@@ -4,76 +4,77 @@ go 1.13
 
 // direct dependencies
 require (
-	github.com/asaskevich/govalidator v0.0.0-20190424111038-f61b66f89f4a
+	github.com/asaskevich/govalidator v0.0.0-20200108200545-475eaeb16496
 	github.com/aws/aws-sdk-go-v2 v0.18.0
 	github.com/blang/semver v3.5.0+incompatible
 	github.com/c9s/goprocinfo v0.0.0-20190309065803-0b2ad9ac246b
-	github.com/census-instrumentation/opencensus-proto v0.2.1 // indirect
 	github.com/cilium/arping v1.0.1-0.20190728065459-c5eaf8d7a710
 	github.com/cilium/ebpf v0.0.0-20191113100448-d9fb101ca1fb
+	github.com/cilium/hubble v0.0.0-20200223004026-fac385525a79
```

Comment:
Why is cilium importing `github.com/cilium/hubble` and hubble importing `github.com/cilium/cilium`?

Reply:
I don't think this is avoidable without significant changes. Hubble needs to import various Cilium type definitions. For example the monitor message types (`TraceNotify` and friends), the reserved identity definitions, the label selector infrastructure, u8proto, etc just to name a few.

The cleanest options I see to avoid this cyclic dependency:
a) Merge the Hubble repository into Cilium (which is something we decided against)
b) Split out all Cilium type definition that Hubble needs of Cilium into it's own go module (e.g. `github.com/cilium/cilium-types`), but given how many different types Hubble needs to import (see above), I don't think there is a clean line we could draw.

I personally think we live with the fact that we do have this cyclic dependency. Go modules seems smart enough to recognized this and seems to do the right thing (i.e. it does not vendor in Cilium as a dependency of Hubble). If we really do not want to rely on go modules, we could also import the Hubble packages e.g. a git submodule or git subtree and pretend they are normal packages of Cilium, but I don't see the advantage over Go modules there.

Reply:
This looks really weird since we are updating random dependencies in Cilium repository for a project that depends on Cilium. For example:

- why is `github.com/fatih/color` being updated to `v1.9.0`? This is a direct dependency of cilium an a indirect dependency of hubble, it's indirect of hubble because cilium is importing it. However, for some reason this dependency is being updated to 1.9.0. in this PR.
```
$ go mod why github.com/fatih/color
# github.com/fatih/color
github.com/cilium/hubble/pkg/cilium
github.com/cilium/cilium/pkg/identity
github.com/cilium/cilium/pkg/option
github.com/cilium/cilium/pkg/color
github.com/fatih/color
```

- In Hubble we are importing cilium rc2 which has, or will have, different direct dependencies than cilium latest, when will we update Hubble with the most recent Cilium version? After that is done shouldn't we also update Hubble in Cilium?

Reply:
The bump of `github.com/fatih/color` is weird indeed. Previously, it might have been explained by Hubble's use of zap (zap also imports that package, but zap is no longer a direct dependency of Hubble), but current Hubble master does not do this.

> In Hubble we are importing cilium rc2 which has, or will have, different direct dependencies than cilium latest, when will we update Hubble with the most recent Cilium version? After that is done shouldn't we also update Hubble in Cilium?

I believe Go Module should never consider the `cilium-rc2` dependencies of Hubble when built inside Cilium. At least that is my understanding of  Go modules [Minimal Version Selection](https://research.swtch.com/vgo-mvs). It replaces `github.com/cilium/cilium 1.7.0-rc2` with `github.com/cilium/cilium master`, therefore ignoring any dependencies that `-rc2` might have had.

The only thing that worries me here that Hubble on its own will always import an old version of Cilium. This means that the Hubble code embedded inside Cilium is never unit tested with the most recent version of Cilium. This to me is a major downside.

Reply:
> It seems really odd that 2 projects are importing each other, can we fix that first before merging this PR?

What exactly is a "fix" here?

> This looks really weird since we are updating random dependencies in Cilium repository for a project that depends on Cilium

That seems like a separate issue apart form circular mod. Cilium mod can probably be expanded in it's locking of all deps to fix this particular issue.

> Why is cilium importing github.com/cilium/hubble and hubble importing github.com/cilium/cilium?

As already mentioned, hubble needs access to data structures defined inside cilium. Resolving this cycle can only be done through intermediate third package that both import, or completely merging the two together. Neither options are particularly lightweight

Reply:
@aanm @tgraf let me know if there is any action i can take here.

Reply:
slightly cleaner commit: https://github.com/cilium/cilium/pull/10238/commits/7765951a4f9bdd2f433739c9b4bb76bba54f035d

this is what i got with `go mod vendor` without doing `go get -u`. @aanm @tgraf let me know it looks ok.

Reply:
updating go.mod in cilium/hubble as well: https://github.com/cilium/hubble/pull/139 this should make them more in sync.

Reply:
updated go.mod: https://github.com/cilium/cilium/pull/10238/commits/e7d675a6eb58bb14a8a997c4535ab7e190d7f4d4

resolving this